Reorder args to come after default args for docker run

antoniobeyah opened the pull request:
Reorder the arguments to allow for user specified arguments to come after all of the hardcoded arguments.

This will allow you to workaround some issues that were created in #63 by allowing you to override items such as entrypoint which may be required for some containers (i.e. systemd must be pid 1).

Before:

```
docker run <default_args> <user_args> <default_args> <image>
```

After:

```
docker run <default_args> <user_args> <image>
```

I added tests, but was unable to validate them due to the error below, seems that it is a [known issue](https://issues.jenkins-ci.org/browse/JENKINS-33962) for docker-machine.

```
[prj] Running shell script
sh: 1: cannot create /var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4687256505953296942/junit1220044917864205478/workspace/prj@tmp/durable-b7f04670/pid: Directory nonexistent
sh: 1: cannot create /var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4687256505953296942/junit1220044917864205478/workspace/prj@tmp/durable-b7f04670/jenkins-log.txt: Directory nonexistent
sh: 1: cannot create /var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4687256505953296942/junit1220044917864205478/workspace/prj@tmp/durable-b7f04670/jenkins-result.txt: Directory nonexistent
[Pipeline] }
```

Output from tests:

```
before:
[prj #1] $ docker run -t -d -u 831285540:20 --entrypoint top -w
/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit6294743115081687909/junit1069789762294073878/workspace/prj
-v
/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit6294743115081687909/junit1069789762294073878/workspace/prj:/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit6294743115081687909/junit1069789762294073878/workspace/prj:rw
-v
/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit6294743115081687909/junit1069789762294073878/workspace/prj@tmp:/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit6294743115081687909/junit1069789762294073878/workspace/prj@tmp:rw
-e ******** -e ******** -e ******** -e ******** -e ******** -e ********
-e ******** -e ******** -e ******** -e ******** -e ******** -e ********
-e ******** -e ******** -e ******** -e ******** --entrypoint cat ubuntu

after:
[prj #1] $ docker run -t -d -u 831285540:20 -w
/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4564385658999144660/junit4862643704887102551/workspace/prj
-v
/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4564385658999144660/junit4862643704887102551/workspace/prj:/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4564385658999144660/junit4862643704887102551/workspace/prj:rw
-v
/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4564385658999144660/junit4862643704887102551/workspace/prj@tmp:/var/folders/p6/66k97xvd7jl2ttk9wgmq2z9hrrrt94/T/junit4564385658999144660/junit4862643704887102551/workspace/prj@tmp:rw
-e ******** -e ******** -e ******** -e ******** -e ******** -e ********
-e ******** -e ******** -e ******** -e ******** -e ******** -e ********
-e ******** -e ******** -e ******** -e ******** --entrypoint cat
--entrypoint top ubuntu
```

cc: @jglick

ten0s commented:
+1

I also need to override the default entrypoint.

jglick commented on src/test/java/org/jenkinsci/plugins/docker/workflow/WithContainerStepTest.java:
```diff
@@ -236,4 +236,38 @@
+                        "  }\n" +
+                        "}", true));
+                WorkflowRun b = story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+                story.j.assertLogContains("cat", b);
```
What is this proving? If you

```groovy
sh 'cat /any/file/at/all'
```

your log will contain

```
+ cat /any/file/at/all
```

jglick commented on src/test/java/org/jenkinsci/plugins/docker/workflow/WithContainerStepTest.java:
```diff
@@ -236,4 +236,38 @@
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "prj");
+                p.setDefinition(new CpsFlowDefinition(
+                        "node {\n" +
+                        "  withDockerContainer(args: '--entrypoint top', image: 'ubuntu') {\n" +
```
And why would you do that to begin with?

labisso replied:
I also have this use case. We have images that use `ENTRYPOINT` to launch background processes via supervisord or runit. These processes are used by the commands within the `inside` block. Since the changes in #63 we've had to stay pinned to 0.7.

jglick commented:
> was unable to validate them due to the error below, seems that it is a known issue for docker-machine

In order to use a remote Docker daemon, you need to running from inside a container, so that `--volumes-from` can be passed to share the workspace.

antoniobeyah commented:
@jglick I have updated the test case to more accurately reflect the need for it.  There are also users reporting issues with the new implementation on the [original jira ticket](https://issues.jenkins-ci.org/browse/JENKINS-37987) and within this pr. This change would allow us to work around the new implementation introduced in #63 .

As I originally stated in the PR, the intent (motivation) is to you to specify your own entrypoint that overrides the hardcoded one.  This is required for containers that rely on systemd.  systemd requires itself to be pid 1, which setting the entrypoint does.

One of the tests had a bug in it that you pointed out, I have fixed it by switching to grep instead of cat.  I am still unable to run the tests locally via IntelliJ, even after switching to Docker for Mac.  If there are instructions/documentation on how to interact with this test harness somewhere I would greatly appreciate a link.  The test was contrived to focus on the code change instead of the specific use case.  I have changed it to be more realistic.

The tests demonstrate two things:
1) When an entrypoint is not specified, the entrypoint is 'cat'
2) When an entrypoint is specified, the entrypoint is the entrypoint that you defined

The tests are written to fail if the string its looking for isn't found, so I removed the test of the log output.  

Additional notes:
- grep will exit 1 if there is no match, 0 otherwise
- /proc/<pid>/cmdline is the command line that the process was run with, this shows up in ps

Please let me know if there is anything else you require for this.  Hopefully this explains the motivation a bit more.

antoniobeyah commented:
@jglick ping.  changes here are pretty simple, its just reordering arguments.

jglick commented:
> specify your own entrypoint that overrides the hardcoded one. This is required for containers that rely on systemd
> […]
> Hopefully this explains the motivation a bit more.

Not really. The test case just demonstrates the effect of the change, but not what you are actually trying to accomplish in a realistic scenario.

In general if the image specifies an “active” entrypoint it is a good hint that using `Image.inside` was never appropriate to begin with, since this feature is designed only for use with “passive” images that merely define a set of binaries and file paths and the like. Possibly you meant to use `Image.run`/`.withRun` instead.

antoniobeyah commented:
@jglick I'm getting a bit confused here about the resistance of merging this change.  #63 broke backwards compatibility without providing a workaround, this change will at least allow you to workaround it without completely changing the structure of your code.  Besides that, the code change is dead simple and has tests.  

At this point I'm not sure if you want anything else or whether or not we should close the PR with the resolution of "use something else/do it different".  If you don't intend on merging it, please close it.  If possible, I would request some clarification on how to achieve similar behavior (executing commands inside a runner container) using this plugin that allows for a custom defined entrypoint.

After reading the docs, it seems like you are suggesting that we would have to do this instead (and I'm not sure whether or not our entrypoint will take precedence in this case either, I haven't tested.  If it calls DockerClient.run, it **won't** work):
```
node {
  git '…'
  docker.image('mysql').withRun('--entrypoint /usr/sbin/init') {c ->
    sh 'docker exec -t ${c.id} command1'
  }
}
```

I arrived at this from the docs (withRun):

> The above simply starts a container running a test MySQL database and runs a regular build while that container is running. Unlike inside, shell steps inside the block are not run inside the container, but they could connect to it using a local TCP port for example.

jglick commented:
> I'm not sure if you want anything else

Yes—a description of an actual, realistic use case that does not currently work. `Docker.Image.inside` is only intended for passive containers which do not require any particular entrypoint. It sounds like you are running some kind of more active container, a service of some kind perhaps, but the details are unspecified.

`withRun` (which merely calls `docker run`) plus `docker exec` as you quoted is a more direct way of running stuff “inside” a container. Generally you would not need to specify an `--entrypoint` if the image already defined a suitable one. Depends on the image and what it is supposed to do.

antoniobeyah commented:
@jglick as discussed in one of the previous comments, the real-world use case is the use of containers that rely on systemd.  The only reason this doesn't work is due to the way this method is ordered.  If you change the way the methods are ordered it will solve my problems and pain of a lot of others.  What are the concerns related to the code changes?

Thanks,
Antonio

jglick commented:
> the use of containers that rely on systemd

OK, so a realistic test case with an example that worked in some earlier release that does not now?

antoniobeyah commented:
@jglick Sure, I'll create a systemd specific test case.

jglick commented:
#78 seems more straightforward. I would encourage you to work with @hogarthj to improve the test case to not be obviously artificial.
